sdk/go/common/resource/config/plaintext.go

```diff
@@ -0,0 +1,187 @@
+type PlaintextType interface {
+	bool | int64 | float64 | string | []Plaintext | map[string]Plaintext
+}
+
+// Plaintext is a single plaintext config value.
+type Plaintext struct {
+	value  any
+	secure bool
+}
+
+// NewPlaintext creates a new plaintext config value.
+func NewPlaintext[T PlaintextType](v T) Plaintext {
+	if m, ok := any(v).(map[string]Plaintext); ok && len(m) == 1 {
+		if _, ok := m["secure"].Value().(string); ok {
+			contract.Failf(errSecureReprReserved.Error())
+		}
+	}
+
+	return Plaintext{value: v}
+}
+
+// NewSecurePlaintext creates a new secure string with the given plaintext.
+func NewSecurePlaintext(plaintext string) Plaintext {
+	return Plaintext{value: plaintext, secure: true}
+}
+
+// Secure returns true if the receiver is a secure string or a composite value that contains a secure string.
+func (c Plaintext) Secure() bool {
+	switch v := c.Value().(type) {
+	case []Plaintext:
+		for _, v := range v {
+			if v.Secure() {
+				return true
+			}
+		}
+		return false
+	case map[string]Plaintext:
+		for _, v := range v {
+			if v.Secure() {
+				return true
+			}
+		}
+		return false
+	case string:
+		return c.secure
+	default:
+		return false
+	}
+}
+
+// Value returns the inner plaintext value.
+func (c Plaintext) Value() any {
+	return c.value
+}
+
+// GoValue returns the inner plaintext value as a plain Go value:
+//
+// - secure strings are mapped to their plaintext
+// - []Plaintext values are mapped to []any values
+// - map[string]Plaintext values are mapped to map[string]any values
+func (c Plaintext) GoValue() any {
+	switch v := c.Value().(type) {
+	case []Plaintext:
+		vs := make([]any, len(v))
+		for i, v := range v {
+			vs[i] = v.GoValue()
+		}
+		return vs
+	case map[string]Plaintext:
+		vs := make(map[string]any, len(v))
+		for k, v := range v {
+			vs[k] = v.GoValue()
+		}
+		return vs
+	default:
+		return v
+	}
+}
+
+// Encrypt converts the receiver as a Value. All secure strings in the result are encrypted using encrypter.
+func (c Plaintext) Encrypt(ctx context.Context, encrypter Encrypter) (Value, error) {
+	obj, err := c.encrypt(ctx, encrypter)
+	if err != nil {
+		return Value{}, err
+	}
+	return obj.marshalValue()
+}
+
+// encrypt converts the receiver to an object. All secure strings in the result are encrypted using encrypter.
+func (c Plaintext) encrypt(ctx context.Context, encrypter Encrypter) (object, error) {
+	switch v := c.Value().(type) {
+	case bool:
+		return newObject(v), nil
+	case int64:
+		return newObject(v), nil
+	case float64:
+		return newObject(v), nil
+	case string:
+		if !c.secure {
+			return newObject(v), nil
+		}
+		ciphertext, err := encrypter.EncryptValue(ctx, v)
+		if err != nil {
+			return object{}, err
+		}
+		return newSecureObject(ciphertext), nil
+	case []Plaintext:
+		vs := make([]object, len(v))
+		for i, v := range v {
+			ev, err := v.encrypt(ctx, encrypter)
+			if err != nil {
+				return object{}, err
```

Comment:
For when there's a deeply nested secret, map and slice should probably add context to the error. Otherwise it'll be super unclear what actually failed.

```go
return object{}, fmt.Errorf("encrypt [%d]: %w", i, err)
// and below,
return object{}, fmt.Errorf("encrypt .%q: %w", k, err)
```

This does have a slight disadvantage of printing the following, but that's better than no contextual information.

```
encrypt [3]: encrypt ."foo": encrypt [5]: encrypt ."bar": great sadness
```

As a follow up in a future change, we can collapse that into something like the following with some bookkeeping.

```
encrypt [3]."foo".[5]."bar": great sadness
// (or something with nicer formatting)
```

We can track the path from root to current item with something like `func (PlainText) encrypt(..., path []string)`:

```go
ev, err := v.encrypt(ctx, encrypter, append(path, strconv.Quote(key))
// ...
return object{}, fmt.Errorf("encrypt %v: %w", strings.Join(path, "."), err)
```